File: src/scripts/lint_command_tiers.py
#!/usr/bin/env python3
"""Lint slash-command frontmatter for the `tier:` key.

Hard-fails CI if any command under .agent-src.uncondensed/commands/
lacks a `tier:` declaration or uses an unknown tier value. The valid
tier set is locked by docs/contracts/command-surface-tiers.md.

Hooked into `task ci` after `task lint-rule-tiers`.

Exit codes:
  0  every command declares a valid tier
  1  one or more commands missing or using an invalid tier
"""
from __future__ import annotations

import sys
from pathlib import Path

QUIET = "--quiet" in sys.argv

REPO = Path(__file__).resolve().parents[2]

# Commands live under every artefact root post-monorepo Phase 4.
sys.path.insert(0, str(Path(__file__).resolve().parent))
from _lib.agent_src import artefact_roots  # noqa: E402

# Post-ADR-051 the command sources live at src/domains/<pack>/**/command.md;
# legacy artefact-root commands/ dirs are kept for older checkouts. A root
# whose commands/ holds only evals (no .md) is not a command source.
COMMANDS_DIRS = [
    root / "commands" for root in artefact_roots()
    if (root / "commands").is_dir()
    and any(p.name != "AGENTS.md" for p in (root / "commands").rglob("*.md"))
]
DOMAINS_DIR = REPO / "src" / "domains"
# Consumer-facing projection — must also carry tier so .augment/commands/
# (which symlinks to dist/agent-src/commands/) renders the tier filter.
COMMANDS_DIR_CONDENSED = REPO / "dist/agent-src" / "commands"

VALID_TIERS = frozenset({"0", "1", "2"})
# ADR-090: `visibility:` is the named source of truth; `tier:` is the
# back-compat integer alias. Both are validated; when both are present they
# MUST agree per this mapping.
VALID_VISIBILITIES = frozenset({"visible", "advanced", "internal"})
TIER_TO_VISIBILITY = {"0": "visible", "1": "advanced", "2": "internal"}


def parse_field(text: str, key: str) -> str | None:
    if not text.startswith("---\n"):
        return None
    end = text.find("\n---\n", 4)
    if end == -1:
        return None
    for line in text[4:end].splitlines():
        if ":" not in line:
            continue
        k, _, v = line.partition(":")
        if k.strip() == key:
            return v.strip().strip('"').strip("'")
    return None


def parse_tier(text: str) -> str | None:
    return parse_field(text, "tier")


def visibility_error(text: str) -> str | None:
    """Validate the visibility field (ADR-090). Returns an error string or None.

    Requires a present + valid visibility, and consistency with the tier alias
    whenever both are declared.
    """
    vis = parse_field(text, "visibility")
    if vis is None:
        return "missing visibility"
    if vis not in VALID_VISIBILITIES:
        return f"invalid visibility '{vis}'"
    tier = parse_field(text, "tier")
    if tier in TIER_TO_VISIBILITY and TIER_TO_VISIBILITY[tier] != vis:
        return f"visibility '{vis}' disagrees with tier '{tier}' (expected '{TIER_TO_VISIBILITY[tier]}')"
    return None


def lint(commands_dir: Path, *, quiet: bool = False) -> int:
    """Lint a commands directory. Returns 0 on success, 1 on failure."""
    if not commands_dir.is_dir():
        print(
            f"lint_command_tiers: no commands dir at {commands_dir}",
            file=sys.stderr,
        )
        return 1

    files = sorted(commands_dir.rglob("*.md"))
    # Sub-AGENTS.md companions are not slash commands.
    commands = [p for p in files if p.name != "AGENTS.md"]

    if not commands:
        print(
            f"lint_command_tiers: no commands found under {commands_dir}",
            file=sys.stderr,
        )
        return 1

    missing: list[str] = []
    invalid: list[tuple[str, str]] = []
    vis_errors: list[tuple[str, str]] = []

    for cmd in commands:
        rel = cmd.relative_to(commands_dir).as_posix()
        text = cmd.read_text(encoding="utf-8")
        tier = parse_tier(text)
        if tier is None:
            missing.append(rel)
        elif tier not in VALID_TIERS:
            invalid.append((rel, tier))
        if (ve := visibility_error(text)) is not None:
            vis_errors.append((rel, ve))

    if missing or invalid or vis_errors:
        print(
            f"❌  lint_command_tiers: {len(missing)} missing tier, "
            f"{len(invalid)} invalid tier, {len(vis_errors)} visibility "
            f"(of {len(commands)} commands)",
            file=sys.stderr,
        )
        for name in missing:
            print(f"    missing tier: {name}", file=sys.stderr)
        for name, tier in invalid:
            print(f"    invalid tier '{tier}': {name}", file=sys.stderr)
        for name, err in vis_errors:
            print(f"    {err}: {name}", file=sys.stderr)
        print(
            f"    valid tiers: {sorted(VALID_TIERS)}; "
            f"valid visibility: {sorted(VALID_VISIBILITIES)}",
            file=sys.stderr,
        )
        print(
            "    contract: docs/contracts/command-surface-tiers.md (ADR-090)",
            file=sys.stderr,
        )
        return 1

    if not quiet:
        print(
            f"✅  lint_command_tiers: {len(commands)} commands, "
            f"all tier + visibility values valid"
        )
    return 0


def lint_domain_sources(*, quiet: bool = False) -> int:
    """Lint src/domains/**/command.md — the post-ADR-051 authoring tree."""
    commands = sorted(DOMAINS_DIR.rglob("command.md"))
    if not commands:
        print(
            f"lint_command_tiers: no command.md found under {DOMAINS_DIR}",
            file=sys.stderr,
        )
        return 1
    missing: list[str] = []
    invalid: list[tuple[str, str]] = []
    vis_errors: list[tuple[str, str]] = []
    for c in commands:
        rel = c.relative_to(REPO).as_posix()
        text = c.read_text(encoding="utf-8")
        tier = parse_tier(text)
        if tier is None:
            missing.append(rel)
        elif tier not in VALID_TIERS:
            invalid.append((rel, tier))
        if (ve := visibility_error(text)) is not None:
            vis_errors.append((rel, ve))
    if missing or invalid or vis_errors:
        print(
            f"❌  lint_command_tiers: {len(missing)} missing tier, "
            f"{len(invalid)} invalid tier, {len(vis_errors)} visibility "
            f"(of {len(commands)} domain commands)",
            file=sys.stderr,
        )
        for name in missing:
            print(f"    missing tier: {name}", file=sys.stderr)
        for name, tier in invalid:
            print(f"    invalid tier '{tier}': {name}", file=sys.stderr)
        for name, err in vis_errors:
            print(f"    {err}: {name}", file=sys.stderr)
        return 1
    if not quiet:
        print(
            f"✅  lint_command_tiers: {len(commands)} domain commands, "
            f"all tier + visibility values valid"
        )
    return 0


def main() -> int:
    if not COMMANDS_DIRS and not DOMAINS_DIR.is_dir():
        print(
            "lint_command_tiers: no commands dir found under any artefact root",
            file=sys.stderr,
        )
        return 1
    rc = 0
    if DOMAINS_DIR.is_dir():
        rc |= lint_domain_sources(quiet=QUIET)
    for commands_dir in COMMANDS_DIRS:
        rc |= lint(commands_dir, quiet=QUIET)
    # The condensed projection is the consumer-facing tree (via the
    # .augment/commands → dist/agent-src/commands symlink). It must also
    # carry tier so the surface stays uniform.
    if COMMANDS_DIR_CONDENSED.is_dir():
        rc |= lint(COMMANDS_DIR_CONDENSED, quiet=QUIET)
    return rc


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/lint_command_tiers.ts
#!/usr/bin/env node
/**
 * Lint slash-command frontmatter for the `tier:` key.
 *
 * TypeScript twin of `src/scripts/lint_command_tiers.py` (ADR-092, Phase 4 /
 * Wave 4b). Mirrors the Python CLI contract exactly: same scan scope, file
 * ordering, finding messages, stdout/stderr split, and exit codes
 * (return-code bitwise-OR accumulation across the domain + per-root +
 * condensed-projection passes).
 *
 * Hard-fails CI if any command lacks a `tier:` declaration or uses an unknown
 * tier value. The valid tier set is locked by
 * docs/contracts/command-surface-tiers.md.
 *
 * Exit codes:
 *   0  every command declares a valid tier
 *   1  one or more commands missing or using an invalid tier
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { artefact_roots } from "./_lib/agent_src.js";

const QUIET = process.argv.includes("--quiet");

// src/scripts/lint_command_tiers.ts → two levels up is the repo root.
const REPO = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..");

function _isDir(p: string): boolean {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}

/** Recursive `*.md` glob, sorted by POSIX string; follows directory symlinks. */
function rglobMd(root: string): string[] {
  const out: string[] = [];
  const walk = (dir: string): void => {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const ent of entries) {
      const full = path.join(dir, ent.name);
      if (ent.name.endsWith(".md")) {
        out.push(full);
      }
      if (ent.isDirectory() || (ent.isSymbolicLink() && _isDir(full))) {
        walk(full);
      }
    }
  };
  walk(root);
  out.sort();
  return out;
}

function relPosix(child: string, base: string): string {
  return path.relative(base, child).split(path.sep).join("/");
}

// Post-ADR-051 the command sources live at src/domains/<pack>/**/command.md;
// legacy artefact-root commands/ dirs are kept for older checkouts. A root
// whose commands/ holds only evals (no .md) is not a command source.
function COMMANDS_DIRS(): string[] {
  return artefact_roots()
    .map((root) => path.join(root, "commands"))
    .filter((d) => {
      if (!_isDir(d)) {
        return false;
      }
      return rglobMd(d).some((p) => path.basename(p) !== "AGENTS.md");
    });
}

const DOMAINS_DIR = path.join(REPO, "src", "domains");
// Consumer-facing projection — must also carry tier so .augment/commands/
// (which symlinks to dist/agent-src/commands/) renders the tier filter.
const COMMANDS_DIR_CONDENSED = path.join(REPO, "dist/agent-src", "commands");

const VALID_TIERS: ReadonlySet<string> = new Set(["0", "1", "2"]);

export function parse_tier(text: string): string | null {
  if (!text.startsWith("---\n")) {
    return null;
  }
  const end = text.indexOf("\n---\n", 4);
  if (end === -1) {
    return null;
  }
  for (const line of text.slice(4, end).split("\n")) {
    if (!line.includes(":")) {
      continue;
    }
    const idx = line.indexOf(":");
    const k = line.slice(0, idx);
    const v = line.slice(idx + 1);
    if (k.trim() === "tier") {
      return stripQuotes(v.trim());
    }
  }
  return null;
}

function stripQuotes(s: string): string {
  return stripChar(stripChar(s, '"'), "'");
}

function stripChar(s: string, ch: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && s[start] === ch) start += 1;
  while (end > start && s[end - 1] === ch) end -= 1;
  return s.slice(start, end);
}

// Mirror Python's `print(sorted(set))` list repr: ['0', '1', ...].
function sortedTiersRepr(): string {
  const items = [...VALID_TIERS].sort();
  return `[${items.map((s) => `'${s}'`).join(", ")}]`;
}

export function lint(commands_dir: string, quiet = false): number {
  if (!_isDir(commands_dir)) {
    process.stderr.write(
      `lint_command_tiers: no commands dir at ${commands_dir}\n`,
    );
    return 1;
  }

  const files = rglobMd(commands_dir);
  // Sub-AGENTS.md companions are not slash commands.
  const commands = files.filter((p) => path.basename(p) !== "AGENTS.md");

  if (commands.length === 0) {
    process.stderr.write(
      `lint_command_tiers: no commands found under ${commands_dir}\n`,
    );
    return 1;
  }

  const missing: string[] = [];
  const invalid: Array<[string, string]> = [];

  for (const cmd of commands) {
    const rel = relPosix(cmd, commands_dir);
    const tier = parse_tier(fs.readFileSync(cmd, "utf-8"));
    if (tier === null) {
      missing.push(rel);
    } else if (!VALID_TIERS.has(tier)) {
      invalid.push([rel, tier]);
    }
  }

  if (missing.length > 0 || invalid.length > 0) {
    process.stderr.write(
      `❌  lint_command_tiers: ${missing.length} missing, ` +
        `${invalid.length} invalid (of ${commands.length} commands)\n`,
    );
    for (const name of missing) {
      process.stderr.write(`    missing tier: ${name}\n`);
    }
    for (const [name, tier] of invalid) {
      process.stderr.write(`    invalid tier '${tier}': ${name}\n`);
    }
    process.stderr.write(`    valid tiers: ${sortedTiersRepr()}\n`);
    process.stderr.write(
      "    contract: docs/contracts/command-surface-tiers.md\n",
    );
    return 1;
  }

  if (!quiet) {
    process.stdout.write(
      `✅  lint_command_tiers: ${commands.length} commands, ` +
        "all tier values valid\n",
    );
  }
  return 0;
}

/** Lint src/domains/**\/command.md — the post-ADR-051 authoring tree. */
export function lint_domain_sources(quiet = false): number {
  const commands = rglobNamed(DOMAINS_DIR, "command.md");
  if (commands.length === 0) {
    process.stderr.write(
      `lint_command_tiers: no command.md found under ${DOMAINS_DIR}\n`,
    );
    return 1;
  }
  const missing: string[] = [];
  const invalid: Array<[string, string]> = [];
  for (const c of commands) {
    const t = parse_tier(fs.readFileSync(c, "utf-8"));
    const rel = relPosix(c, REPO);
    if (t === null) {
      missing.push(rel);
    } else if (!VALID_TIERS.has(t)) {
      invalid.push([rel, t]);
    }
  }
  if (missing.length > 0 || invalid.length > 0) {
    process.stderr.write(
      `❌  lint_command_tiers: ${missing.length} missing, ` +
        `${invalid.length} invalid (of ${commands.length} domain commands)\n`,
    );
    for (const name of missing) {
      process.stderr.write(`    missing tier: ${name}\n`);
    }
    for (const [name, tier] of invalid) {
      process.stderr.write(`    invalid tier '${tier}': ${name}\n`);
    }
    return 1;
  }
  if (!quiet) {
    process.stdout.write(
      `✅  lint_command_tiers: ${commands.length} domain commands, ` +
        "all tier values valid\n",
    );
  }
  return 0;
}

/** Recursive glob for an exact filename, sorted. */
function rglobNamed(root: string, name: string): string[] {
  const out: string[] = [];
  const walk = (dir: string): void => {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const ent of entries) {
      const full = path.join(dir, ent.name);
      if (ent.name === name) {
        out.push(full);
      }
      if (ent.isDirectory() || (ent.isSymbolicLink() && _isDir(full))) {
        walk(full);
      }
    }
  };
  walk(root);
  out.sort();
  return out;
}

export function main(): number {
  const commandsDirs = COMMANDS_DIRS();
  if (commandsDirs.length === 0 && !_isDir(DOMAINS_DIR)) {
    process.stderr.write(
      "lint_command_tiers: no commands dir found under any artefact root\n",
    );
    return 1;
  }
  let rc = 0;
  if (_isDir(DOMAINS_DIR)) {
    rc |= lint_domain_sources(QUIET);
  }
  for (const commandsDir of commandsDirs) {
    rc |= lint(commandsDir, QUIET);
  }
  // The condensed projection is the consumer-facing tree (via the
  // .augment/commands → dist/agent-src/commands symlink). It must also
  // carry tier so the surface stays uniform.
  if (_isDir(COMMANDS_DIR_CONDENSED)) {
    rc |= lint(COMMANDS_DIR_CONDENSED, QUIET);
  }
  return rc;
}

const isCliEntry =
  process.argv[1] !== undefined &&
  import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (isCliEntry) {
  process.exit(main());
}
